File: models/cv.go
package models

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"html/template"
	"os"
	"strconv"
	"time"

	"github.com/mjarkk/jsonschema"
	"github.com/script-development/RT-CV/helpers/jsonHelpers"
)

// CV contains all information that belongs to a curriculum vitae
// TODO check the json removed fields if we actually should use them
type CV struct {
	Title           string                       `json:"-"` // Not supported yet
	ReferenceNumber string                       `json:"referenceNumber" description:"The reference number (ID) of this CV on the site it was scraped from. We use this to track duplicated CVs"`
	Link            *string                      `json:"link" description:"A link to the page on the site where this cv was found"`
	CreatedAt       *jsonHelpers.RFC3339Nano     `json:"createdAt,omitempty"`
	LastChanged     *jsonHelpers.RFC3339Nano     `json:"lastChanged,omitempty"`
	Educations      []Education                  `json:"educations,omitempty"`
	WorkExperiences []WorkExperience             `json:"workExperiences,omitempty"`
	PreferredJobs   []string                     `json:"preferredJobs,omitempty"`
	Languages       []Language                   `json:"languages,omitempty"`
	Competences     []Competence                 `json:"-"` // Not supported yet
	Interests       []Interest                   `json:"-"` // Not supported yet
	PersonalDetails PersonalDetails              `json:"personalDetails" jsonSchema:"notRequired"`
	Presentation    string                       `json:"presentation" description:"How this person would present him/her self"`
	DriversLicenses []jsonHelpers.DriversLicense `json:"driversLicenses,omitempty"`
}

// Education is something a user has followed
type Education struct {
	Is uint8 `json:"is" description:"What kind of education is this?: 0: Unknown, 1: Education, 2: Course"`

	Name        string `json:"name"`
	Description string `json:"description"`
	Institute   string `json:"institute"`
	// TODO find difference between isCompleted and hasdiploma
	IsCompleted bool                     `json:"isCompleted"`
	HasDiploma  bool                     `json:"hasDiploma"`
	StartDate   *jsonHelpers.RFC3339Nano `json:"startDate"`
	EndDate     *jsonHelpers.RFC3339Nano `json:"endDate"`
}

// WorkExperience is experience in work
type WorkExperience struct {
	Description       string                   `json:"description"`
	Profession        string                   `json:"profession"`
	StartDate         *jsonHelpers.RFC3339Nano `json:"startDate"`
	EndDate           *jsonHelpers.RFC3339Nano `json:"endDate"`
	StillEmployed     bool                     `json:"stillEmployed"`
	Employer          string                   `json:"employer"`
	WeeklyHoursWorked int                      `json:"weeklyHoursWorked"`
}

// LanguageLevel is something that i'm not sure what it is
type LanguageLevel uint

// The lanague levels available
const (
	LanguageLevelUnknown LanguageLevel = iota
	LanguageLevelReasonable
	LanguageLevelGood
	LanguageLevelExcellent
)

func (ll LanguageLevel) String() string {
	switch ll {
	case LanguageLevelReasonable:
		return "Redelijk"
	case LanguageLevelGood:
		return "Goed"
	case LanguageLevelExcellent:
		return "Uitstekend"
	default:
		return "Onbekend"
	}
}

const langLevelDescription = `0. Unknown
1. Reasonable
2. Good
3. Excellent`

// Valid returns weather the language level is valid
func (ll LanguageLevel) Valid() bool {
	return ll >= LanguageLevelUnknown && ll <= LanguageLevelExcellent
}

func (ll LanguageLevel) asjson() json.RawMessage {
	return []byte(strconv.FormatUint(uint64(ll), 10))
}

// JSONSchemaDescribe implements schema.Describe
func (LanguageLevel) JSONSchemaDescribe() jsonschema.Property {
	return jsonschema.Property{
		Title:       "Language level",
		Description: langLevelDescription,
		Type:        jsonschema.PropertyTypeNumber,
		Enum: []json.RawMessage{
			LanguageLevelUnknown.asjson(),
			LanguageLevelReasonable.asjson(),
			LanguageLevelGood.asjson(),
			LanguageLevelExcellent.asjson(),
		},
	}
}

// Language is a language a user can speak
type Language struct {
	Name         string        `json:"name"`
	LevelSpoken  LanguageLevel `json:"levelSpoken"`
	LevelWritten LanguageLevel `json:"levelWritten"`
}

// Competence is an activity a user is "good" at
type Competence struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// Interest contains a job the user is interested in
type Interest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// PersonalDetails contains personal info
type PersonalDetails struct {
	Initials          string                   `json:"initials,omitempty" jsonSchema:"notRequired"`
	FirstName         string                   `json:"firstName,omitempty" jsonSchema:"notRequired"`
	SurNamePrefix     string                   `json:"surNamePrefix,omitempty" jsonSchema:"notRequired"`
	SurName           string                   `json:"surName,omitempty" jsonSchema:"notRequired"`
	DateOfBirth       *jsonHelpers.RFC3339Nano `json:"dob,omitempty" jsonSchema:"notRequired"`
	Gender            string                   `json:"gender,omitempty" jsonSchema:"notRequired"`
	StreetName        string                   `json:"streetName,omitempty" jsonSchema:"notRequired"`
	HouseNumber       string                   `json:"houseNumber,omitempty" jsonSchema:"notRequired"`
	HouseNumberSuffix string                   `json:"houseNumberSuffix,omitempty" jsonSchema:"notRequired"`
	Zip               string                   `json:"zip,omitempty" jsonSchema:"notRequired"`
	City              string                   `json:"city,omitempty" jsonSchema:"notRequired"`
	Country           string                   `json:"country,omitempty" jsonSchema:"notRequired"`
	PhoneNumber       *jsonHelpers.PhoneNumber `json:"phoneNumber,omitempty" jsonSchema:"notRequired"`
	Email             string                   `json:"email,omitempty" jsonSchema:"notRequired"`
}

func getTemplateFromFile(funcs template.FuncMap, filename string) (*template.Template, error) {
	tmpl, err := template.New(filename).Funcs(funcs).ParseFiles("./assets/" + filename)
	if err != nil {
		if !os.IsNotExist(err) {
			return nil, err
		}

		// For testing perposes
		tmpl, err = template.New(filename).Funcs(funcs).ParseFiles("../assets/" + filename)
		if err != nil {
			return nil, err
		}
	}
	return tmpl, nil
}

// GetEmailHTML generates a HTML document that is used as email body
func (cv *CV) GetEmailHTML(profile Profile, matchText, domain string) (*bytes.Buffer, error) {
	tmpl, err := getTemplateFromFile(template.FuncMap{}, "email-template.html")
	if err != nil {
		return nil, err
	}

	input := struct {
		Profile   Profile
		Cv        *CV
		MatchText string
		LogoURL   string
		Domain    string

		// The normal `Profile.ID.String()`` is more of a debug value than a real id value so we add the hex to this field
		ProfileIDHex string
	}{
		Profile:      profile,
		ProfileIDHex: profile.ID.Hex(),
		Cv:           cv,
		MatchText:    matchText,
		LogoURL:      os.Getenv("EMAIL_LOGO_URL"),
		Domain:       domain,
	}

	buff := bytes.NewBuffer(nil)
	err = tmpl.Execute(buff, input)
	return buff, err
}

// Validate validates the cv and returns an error if it's not valid
func (cv *CV) Validate() error {
	// TODO: Needs more validation

	if cv.ReferenceNumber == "" {
		return errors.New("referenceNumber must be set")
	}

	now := time.Now()
	tomorrow := now.Add(time.Hour * 24)

	if cv.CreatedAt != nil && cv.CreatedAt.Time().After(tomorrow) {
		return errors.New("createdAt can't be in the future")
	}
	if cv.LastChanged != nil && cv.LastChanged.Time().After(tomorrow) {
		return errors.New("lastChanged can't be in the future")
	}
	if cv.PersonalDetails.DateOfBirth != nil && cv.PersonalDetails.DateOfBirth.Time().After(now.AddDate(-13, 0, 0)) {
		return errors.New("you need to be at least 13 years old to work")
	}

	for idx, lang := range cv.Languages {
		if !lang.LevelSpoken.Valid() {
			return fmt.Errorf("languages.%d.levelSpoken is invalid", idx)
		}
		if !lang.LevelWritten.Valid() {
			return fmt.Errorf("languages.%d.levelWritten is invalid", idx)
		}
	}

	return nil
}

// FullName returns the full name of the cv
func (cv *CV) FullName() string {
	details := cv.PersonalDetails

	res := details.FirstName
	if details.SurName == "" {
		return res
	}

	if details.SurNamePrefix == "" {
		return res + " " + details.SurName
	}

	return res + " " + details.SurNamePrefix + " " + details.SurName
}

// ExampleCV is a cv that can be used for demonstrative purposes
func ExampleCV() *CV {
	link := "https://website-this-cv-came-from.ninja/person/4455"
	now := jsonHelpers.RFC3339Nano(time.Now()).ToPtr()
	return &CV{
		Title:           "Pilot with experience in farming simulator 2020",
		ReferenceNumber: "4455-PIETER",
		Link:            &link,
		CreatedAt:       now,
		LastChanged:     now,

		Educations: []Education{{
			Name:        "Education name",
			Description: "Education description",
			Institute:   "Institute name",
			IsCompleted: true,
			HasDiploma:  true,
			StartDate:   now,
			EndDate:     now,
		}},
		WorkExperiences: []WorkExperience{{
			Description:       "WorkExperience description",
			Profession:        "hitman",
			StartDate:         now,
			EndDate:           now,
			StillEmployed:     true,
			Employer:          "Bond.. James bond",
			WeeklyHoursWorked: 60,
		}},
		PreferredJobs: []string{"hitman"},
		Languages: []Language{{
			Name:         "Language name",
			LevelSpoken:  LanguageLevelExcellent,
			LevelWritten: LanguageLevelGood,
		}},
		Competences: []Competence{{
			Name:        "Competence name",
			Description: "Competence description",
		}},
		Interests: []Interest{{
			Name:        "Interest name",
			Description: "Interest description",
		}},
		Presentation: "I'm a stronk pilot. i'm also very good in farming simulator 2020. Hire Me!",
		DriversLicenses: []jsonHelpers.DriversLicense{
			jsonHelpers.NewDriversLicense("A"),
			jsonHelpers.NewDriversLicense("B"),
			jsonHelpers.NewDriversLicense("C"),
		},
		PersonalDetails: PersonalDetails{
			Initials:          "P.S.",
			FirstName:         "Pietter",
			SurNamePrefix:     "Ven ther",
			SurName:           "Steen",
			DateOfBirth:       now,
			Gender:            "Male",
			StreetName:        "Streetname abc",
			HouseNumber:       "33",
			HouseNumberSuffix: "b",
			Zip:               "9779AB",
			City:              "Groningen",
			Country:           "Netherlands",
			PhoneNumber:       &jsonHelpers.PhoneNumber{IsLocal: true, Number: 611223344},
			Email:             "p.steen@very-smart-people.com",
		},
	}
}

File: pdf_generator/bin/cv.dart
class CV {
  static example() {
    return CV(
      referenceNumber: "d1b77757-144a-4e67-871d-3a3c0f910c9b",
      createdAt: DateTime.now(),
      lastChanged: DateTime.now(),
      personalDetails: PersonalDetails(
        firstName: "Meneer",
        surName: "banaanmans",
        email: "some-email@example.com",
        phoneNumber: "06 11111111",
        dob: DateTime.now(),
        zip: "1234AB",
        streetName: "Straatnaam",
        houseNumber: "15",
        houseNumberSuffix: "a",
        city: "Groningen",
        country: "Nederland",
      ),
      workExperiences: [
        WorkExperience(
          profession: "Gangster",
          employer: "Gangsters B.V.",
          startDate: DateTime.utc(2000, 1, 1),
          endDate: DateTime.utc(2010, 1, 1),
          description: "Doing that gangster shit you know",
        ),
        WorkExperience(
          profession: "Bel meneer",
          employer: "De bel mensen",
          startDate: DateTime.utc(1995, 2, 1),
          endDate: DateTime.utc(1998, 5, 1),
          description:
              "Je weet wel die hinderlijke mensen die je opbellen met onzin waar je helemaal geen belang bij hebt",
        ),
      ],
      educations: [
        Education(
          "HBO",
          institute: "Hogeschool",
          startDate: DateTime.utc(2000, 1, 1),
          endDate: DateTime.utc(2010, 1, 1),
        ),
        Education(
          "MBO",
          institute: "Some school",
          startDate: DateTime.utc(1995, 2, 1),
          endDate: DateTime.utc(1998, 5, 1),
        ),
      ],
      courses: [
        Course(
          "beeing smart",
          institute: "Hogeschool",
          startDate: DateTime.utc(2010, 1, 1),
          endDate: DateTime.utc(2011, 1, 1),
        ),
        Course(
          "TV host voor dummies",
          institute: "Hogeschool",
          startDate: DateTime.utc(2010, 1, 1),
          // endDate: DateTime.utc(2011, 1, 1),
        ),
      ],
      driversLicenses: ["B", "C", "D"],
      languages: [
        Language(
          "Nederlands",
          LanguageLevel.good,
          LanguageLevel.excellent,
        ),
        Language(
          "Duits",
          LanguageLevel.reasonable,
          LanguageLevel.good,
        ),
      ],
    );
  }

  CV.fromJson(Map<String, dynamic> json) {
    referenceNumber = json['referenceNumber'];
    createdAt = jsonParseDate(json['createdAt']);
    lastChanged = jsonParseDate(json['lastChanged']);
    educations = json['educations']
        ?.map((entry) => Education.fromJson(entry))
        ?.toList()
        ?.cast<Education>();
    courses = json['courses']
        ?.map((entry) => Course.fromJson(entry))
        ?.toList()
        ?.cast<Course>();
    workExperiences = json['workExperiences']
        ?.map((entry) => WorkExperience.fromJson(entry))
        ?.toList()
        ?.cast<WorkExperience>();
    preferredJobs = json['preferredJobs']?.cast<String>();
    languages = json['languages']
        ?.map((entry) => Language.fromJson(entry))
        ?.toList()
        ?.cast<Language>();
    personalDetails = PersonalDetails.fromJson(json['personalDetails']);
    driversLicenses = json['driversLicenses']?.cast<String>();
  }

  CV({
    required this.referenceNumber,
    this.createdAt,
    this.lastChanged,
    this.educations,
    this.courses,
    this.workExperiences,
    this.preferredJobs,
    this.languages,
    required this.personalDetails,
    this.driversLicenses,
  });

  late final String referenceNumber;
  late final DateTime? createdAt;
  late final DateTime? lastChanged;
  late final List<Education>? educations;
  late final List<Course>? courses;
  late final List<WorkExperience>? workExperiences;
  late final List<String>? preferredJobs;
  late final List<Language>? languages;
  late final PersonalDetails personalDetails;
  late final List<String>? driversLicenses;
}

class Education {
  Education.fromJson(Map<String, dynamic> json) {
    name = json['name'];
    description = json['description'];
    institute = json['institute'];
    isCompleted = json['isCompleted'];
    hasDiploma = json['hasDiploma'];
    startDate = jsonParseDate(json['startDate']);
    endDate = jsonParseDate(json['endDate']);
  }

  Education(
    this.name, {
    this.description,
    this.institute,
    this.isCompleted,
    this.hasDiploma,
    this.startDate,
    this.endDate,
  });

  late final String name;
  late final String? description;
  late final String? institute;
  late final bool? isCompleted;
  late final bool? hasDiploma;
  late final DateTime? startDate;
  late final DateTime? endDate;
}

class Course {
  Course.fromJson(Map<String, dynamic> json) {
    name = json['name'];
    institute = json['institute'];
    startDate = jsonParseDate(json['startDate']);
    endDate = jsonParseDate(json['endDate']);
    isCompleted = json['isCompleted'];
    description = json['description'];
  }

  Course(
    this.name, {
    this.institute,
    this.startDate,
    this.endDate,
    this.isCompleted,
    this.description,
  });

  late final String name;
  late final String? institute;
  late final DateTime? startDate;
  late final DateTime? endDate;
  late final bool? isCompleted;
  late final String? description;
}

class WorkExperience {
  WorkExperience.fromJson(Map<String, dynamic> json) {
    profession = json['profession'];
    description = json['description'];
    startDate = jsonParseDate(json['startDate']);
    endDate = jsonParseDate(json['endDate']);
    stillEmployed = json['stillEmployed'];
    employer = json['employer'];
    weeklyHoursWorked = json['weeklyHoursWorked'];
  }

  WorkExperience({
    this.profession,
    this.description,
    this.startDate,
    this.endDate,
    this.stillEmployed,
    this.employer,
    this.weeklyHoursWorked,
  });

  late final String? profession;
  late final String? description;
  late final DateTime? startDate;
  late final DateTime? endDate;
  late final bool? stillEmployed;
  late final String? employer;
  late final int? weeklyHoursWorked;
}

enum LanguageLevel {
  unknown,
  reasonable,
  good,
  excellent,
}

Map<LanguageLevel, String> humanLanguageLevel = {
  LanguageLevel.unknown: "Onbekend",
  LanguageLevel.reasonable: "Redelijk",
  LanguageLevel.good: "Goed",
  LanguageLevel.excellent: "Uitstekend",
};

Map<LanguageLevel, int> languageLevelToNr = {
  LanguageLevel.unknown: 0,
  LanguageLevel.reasonable: 1,
  LanguageLevel.good: 2,
  LanguageLevel.excellent: 3,
};

Map<int, LanguageLevel> nrToLanguageLevel = {
  0: LanguageLevel.unknown,
  1: LanguageLevel.reasonable,
  2: LanguageLevel.good,
  3: LanguageLevel.excellent,
};

const int maxLanguageLevelNr = 3;

class Language {
  Language.fromJson(Map<String, dynamic> json) {
    name = json['name'];
    levelSpoken =
        nrToLanguageLevel[json['levelSpoken']] ?? LanguageLevel.unknown;
    levelWritten =
        nrToLanguageLevel[json['levelWritten']] ?? LanguageLevel.unknown;
  }

  Language(this.name, this.levelSpoken, this.levelWritten);

  late final String name;
  late final LanguageLevel levelSpoken;
  late final LanguageLevel levelWritten;
}

class Competence {
  Competence(this.name, {this.description});

  String name;
  String? description;
}

class Interest {
  Interest(this.name, {this.description});

  String name;
  String? description;
}

class PersonalDetails {
  PersonalDetails({
    this.initials,
    this.firstName,
    this.surNamePrefix,
    this.surName,
    this.dob,
    this.gender,
    this.streetName,
    this.houseNumber,
    this.houseNumberSuffix,
    this.zip,
    this.city,
    this.country,
    this.phoneNumber,
    this.email,
  });

  PersonalDetails.fromJson(Map<String, dynamic> json) {
    initials = json['initials'];
    firstName = json['firstName'];
    surNamePrefix = json['surNamePrefix'];
    surName = json['surName'];
    dob = jsonParseDate(json['dob']);
    gender = json['gender'];
    streetName = json['streetName'];
    houseNumber = json['houseNumber'];
    houseNumberSuffix = json['houseNumberSuffix'];
    zip = json['zip'];
    city = json['city'];
    country = json['country'];
    phoneNumber = json['phoneNumber'];
    email = json['email'];
  }

  late final String? initials;
  late final String? firstName;
  late final String? surNamePrefix;
  late final String? surName;
  late final DateTime? dob;
  late final String? gender;
  late final String? streetName;
  late final String? houseNumber;
  late final String? houseNumberSuffix;
  late final String? zip;
  late final String? city;
  late final String? country;
  late final String? phoneNumber;
  late final String? email;

  bool get hasAddress =>
      streetName != null && houseNumber != null && city != null;

  String get fullName {
    String resp = this.firstName ?? '';

    if (this.surName != null) {
      if (resp.isNotEmpty) resp += " ";

      resp += (this.surNamePrefix != null ? this.surNamePrefix! + " " : '') +
          this.surName!;
    }

    return resp;
  }
}

DateTime? jsonParseDate(dynamic input) {
  if (input == null) return null;
  return DateTime.parse(input);
}
